File: studio/repositories/analytics_repo.py
"""Analytics event persistence."""

import time
import uuid

from studio import db
from studio.models import Analytics, to_doc


def track(project_id, event_type, *, metadata=None):
    row = Analytics(
        eventId=uuid.uuid4().hex[:12],
        projectId=project_id,
        eventType=event_type,
        metadata_=metadata or {},
        ts=time.time(),
    )
    with db.session() as s:
        s.add(row)
        s.commit()
        return to_doc(row)


def list_for_project(project_id):
    with db.session() as s:
        rows = (
            s.query(Analytics)
            .filter_by(projectId=project_id)
            .order_by(Analytics.ts.desc())
            .all()
        )
        return [to_doc(r) for r in rows]


def summarize(project_id):
    events = list_for_project(project_id)
    counts = {}
    for event in events:
        key = event.get("eventType", "unknown")
        counts[key] = counts.get(key, 0) + 1
    return {"projectId": project_id, "counts": counts, "total": len(events)}

File: src/server/repos/analyticsRepo.ts
import { Prisma } from "@/generated/prisma/client";
import { prisma } from "../db";
import { toDoc } from "./doc";

export async function track(
  projectId: string,
  eventType: string,
  metadata: Record<string, unknown> = {},
): Promise<Record<string, unknown>> {
  const row = await prisma().builderAnalytic.create({
    data: {
      eventId: randomId(),
      projectId,
      eventType,
      metadata: metadata as Prisma.InputJsonValue,
      ts: Date.now() / 1000,
    },
  });
  return toDoc(row) as Record<string, unknown>;
}

export async function listForProject(projectId: string): Promise<Record<string, unknown>[]> {
  const rows = await prisma().builderAnalytic.findMany({
    where: { projectId },
    orderBy: { ts: "desc" },
  });
  return rows.map((row) => toDoc(row) as Record<string, unknown>);
}

export async function summarize(projectId: string): Promise<Record<string, unknown>> {
  const events = await listForProject(projectId);
  const counts: Record<string, number> = {};
  for (const event of events) {
    const key = (event.eventType as string | undefined) ?? "unknown";
    counts[key] = (counts[key] ?? 0) + 1;
  }
  return { projectId, counts, total: events.length };
}

function randomId(): string {
  const buf = new Uint8Array(6);
  crypto.getRandomValues(buf);
  return Array.from(buf, (b) => b.toString(16).padStart(2, "0")).join("");
}
